### DISCUSSION TITLE
Policy Configuration Issue

### DISCUSSION THREAD

**Discussion by programmerarunkumar** [ImageMagick/ImageMagick#7207] [Help]
### ImageMagick version

7.1.0-60

### Operating system, version and so on

CentOS,7

### Description

[My Code](https://drive.google.com/file/d/1C_L5mHZGGNjkv8s0I9SStV_4XNkHt4bD/view?usp=sharing)

**ISSUE-1**
I have configured the disk limit in policy.xml. IM is not using the configured disk size. it use more than that.

In code im getting the different value . I'm using MagickGetResourceLimit(DiskResource) to get the disk size in the code. But im getting the disk value configured in policy.xml in identify -list resource command

When Im setting the disk limit in the code it is working properly.
Used MagickSetResourceLimit(DiskResource, 2ll * 1024 * 1024 * 1024) to set the disk limit to 2GB

**ISSUE-2**
I have configured the maximum parallel threads to 3 in policy.xml. But in the command line and in the code im getting 1 for the maximum parallel threads.

<img width="182" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/58946415/98558d9a-454d-4e6a-9f64-0d8b6c6f3e08">

**Comment by snibgo** [comment]
> I have configured the disk limit in policy.xml.

Please show your policy.xml.

> IM is not using the configured disk size. it use more than that.

How do you know? The disk limit in policy.xml is only for the pixel cache. IM will use the disk for other purposes, such as writing images to files.

> I have configured the maximum parallel threads to 3 in policy.xml. But in the command line and in the code im getting 1 for the maximum parallel threads.

Perhaps your CPU or O/S supports only one thread.

**Reply by programmerarunkumar** [reply]
> Please show your policy.xml.

> How do you know?

I have attached my policy.xml and my code in my first message Named as My Code.kindly check that

**Comment by snibgo** [comment]
Your program compiles with warnings:
```
snibgo\wresize.c: In function 'main':
snibgo\wresize.c:109:30: warning: passing argument 1 of 'sprintf' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  109 |             sprintf(images[i].resizedImage, "resized_image%d.png", imageNumber+i);
      |                     ~~~~~~~~~^~~~~~~~~~~~~
In file included from snibgo\wresize.c:1:
/usr/include/stdio.h:244:18: note: expected 'char * restrict' but argument is of type 'const char *'
  244 | int     sprintf (char *__restrict, const char *__restrict, ...)
      |                  ^
snibgo\wresize.c:118:27: warning: passing argument 1 of 'free' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  118 |             free(images[i].resizedImage);
      |                  ~~~~~~~~~^~~~~~~~~~~~~
In file included from snibgo\wresize.c:2:
/usr/include/stdlib.h:94:15: note: expected 'void *' but argument is of type 'const char *'
   94 | void    free (void *) _NOTHROW;
      |               ^~~~~~
snibgo\wresize.c:100:13: warning: variable 'threadNum' set but not used [-Wunused-but-set-variable]
  100 |         int threadNum[noOfThread];
      |             ^~~~~~~~~
snibgo\wresize.c: In function 'resize':
snibgo\wresize.c:81:1: warning: control reaches end of non-void function [-Wreturn-type]
   81 | }
      | ^
```

When I run the program, I can see no evidence that:

> IM is not using the configured disk size. it use more than that.

**Comment by programmerarunkumar** [comment]
I have configured the disk limit as 2GB in policy.xml

I have comment the below line in my code. I have got 9223372036854775807 as disk limit
`MagickSetResourceLimit(DiskResource, 2ll * 1024 * 1024 * 1024)`
when i uncomment the above line. I have got 2147483648 as disk limit

**Why the disk limit configured in the policy.xml is not working ?**


[Attachment](https://drive.google.com/file/d/11EdJNCPvdjbQ87DEG1ebNlONEzgMWI1t/view?usp=sharing)
I have add my code, policy.xml, image, and the output screenshot when i have configured the disk limit and when I haven't configured the disk limit in my code

**Comment by snibgo** [comment]
>     printf("Current disk limit: %llu \n", MagickGetResourceLimit(DiskResource));

I don't know if `%llu` is correct for your build of IM. You should use `MagickSizeFormat`, which is defined in magick-type.h.

**Reply by programmerarunkumar** [reply]
okay I will try but

**KFYI**
1.If I haven't configured the disk limit in the code and having 2GB for disk in policy.xml I can see the disk usage is more than 2 GB in the Temp Folder

2.If I have configured the disk limit as 2GB in the code and having 2GB for disk in policy.xml I don't see the disk usage more than 2GB in the Temp Folder. If it is exceeded im properly getting the exception


You can try the above case from my code by commenting and not commenting the below line in my code
`MagickSetResourceLimit(DiskResource, 2ll * 1024 * 1024 * 1024)`

You can run my code as above said and check the disk usage in the temp folder

**Comment by dlemstra** [comment]
I don't see you making sure that the `policy.xml` is read from the current folder? Just putting it there will not cause it to be used.

**Reply by programmerarunkumar** [reply]
I have shared  my policy.xml for you reference 
It will be in etc/ImageMagick-7 folder only. I haven't changed it location.

**Comment by dlemstra** [comment]
Attaching the code here for future reference instead.

[ImageMagickPolicyIssue.zip](https://github.com/ImageMagick/ImageMagick/files/15312757/ImageMagickPolicyIssue.zip)

**Comment by programmerarunkumar** [comment]
Team, do we need any inputs from me ?
Kindly help me to resolve this issue. Thanks in Advance
